src/read/wasm.rs

```diff
@@ -54,23 +63,229 @@ impl<'data> WasmFile<'data> {
 
+        let mut main_file_symbol = Some(Symbol {
+            name: None,
+            address: 0,
+            size: 0,
+            kind: SymbolKind::File,
+            section: SymbolSection::None,
+            weak: false,
+            scope: SymbolScope::Compilation,
+            flags: SymbolFlags::None,
+        });
+
+        let mut imported_funcs_count = 0;
+        let mut local_func_kinds = Vec::new();
+        let mut entry_func_id = None;
+
         for section in module {
-            let section = section.map_err(|_| "Invalid section header")?;
+            let section = section.map_err(|_| "Couldn't read section header")?;
 
             match section.code {
-                wp::SectionCode::Custom { kind, name } => {
-                    if kind == wp::CustomSectionKind::Name {
-                        file.names_data = Some(section.range().slice(data));
-                    } else if name.starts_with(".debug_") {
-                        file.has_debug_symbols = true;
+                wp::SectionCode::Import => {
+                    let mut last_module_name = None;
+
+                    for import in section
+                        .get_import_section_reader()
+                        .map_err(|_| "Couldn't read header of the import section")?
+                    {
+                        let import = import.map_err(|_| "Couldn't read an import item")?;
+                        let module_name = Some(import.module);
+
+                        if last_module_name != module_name {
+                            file.symbols.push(Symbol {
+                                name: module_name,
+                                address: 0,
+                                size: 0,
+                                kind: SymbolKind::File,
+                                section: SymbolSection::None,
+                                weak: false,
+                                scope: SymbolScope::Dynamic,
+                                flags: SymbolFlags::None,
+                            });
+                            last_module_name = module_name;
+                        }
+
+                        let kind = match import.ty {
+                            wp::ImportSectionEntryType::Function(_) => {
+                                imported_funcs_count += 1;
+                                SymbolKind::Text
+                            }
+                            wp::ImportSectionEntryType::Table(_)
+                            | wp::ImportSectionEntryType::Memory(_)
+                            | wp::ImportSectionEntryType::Global(_) => SymbolKind::Data,
+                        };
+
+                        file.symbols.push(Symbol {
+                            name: Some(import.field),
+                            address: 0,
+                            size: 0,
+                            kind,
+                            section: SymbolSection::Section(SectionIndex(SECTION_IMPORT)),
```

Comment:
Imports should be `SymbolSection::Undefined`.

Reply:
Why? I know they're for objects on other platforms, but I think it's because other platforms don't encode type of the import, while in Wasm we get more information.

Reply:
The meaning of this field is to indicate which section contains the object that the symbol is describing, not where the symbol struct is placed. We really don't care where the symbol struct is placed.

If it makes it clearer, if we did the same thing you're doing here for ELF then all symbols would specify the section containing the symbol table, which I think is clearly incorrect.

Reply:
Hmm it feels like duplication then.

In this implementation `kind` field clearly describes whether the symbol points to a function, data, or something else, whereas `section` describes where the symbol itself is placed.

If I change `section` to point to code section, data, etc., then I only get duplication, and it becomes harder to distinguish imported symbols from e.g. exported ones, because both now would have `SymbolScope::Dynamic`, `SymbolKind::Text` and `SymbolSection::Section(SectionIndex(SECTION_CODE))`.

We still need to somehow preserve the information about the section in which symbol itself is defined, and for that `section` field seems like the most appropriate place...

Reply:
This must use the same semantics as the other implementations, or there is no point having a unified API. If that is insufficient, then we'll need to extend the API, but I don't yet see any need for that.

You can still distinguish between local/import/export:
- locals are `SymbolScope::Compilation` and `SymbolSection::Section(SectionIndex(SECTION_CODE))` (or `SECTION_DATA`?)
- exports are `SymbolScope::Dynamic` and `SymbolSection::Section(SectionIndex(SECTION_CODE))`
- imports are `SymbolScope::Dynamic` and `SymbolSection::Undefined`

`SymbolKind` is a little bit redundant with `SymbolSection` (and I'm open to suggestions for improving that. e.g. for Mach-O `SymbolKind` is derived directly from the `SectionKind` and provides no additional information).

Reply:
That model mostly seems reasonable. But one question: how would non-function imports look like? E.g. how would you distinguish a memory import from a function one?

Overall agreed regarding redundancy of `SymbolSection`+`SymbolKind`, but in this case it seems that only `SymbolKind` would currently help?

If so, I'd be open to merging them in a single enum if you are open to changing API that way.

Reply:
> That model mostly seems reasonable. But one question: how would non-function imports look like? E.g. how would you distinguish a memory import from a function one?

Yeah that's one situation where `SymbolKind` isn't redundant; you use it to distinguish, because there is no section that you can use. But not all file formats provide this information and they end up with `SymbolKind::Unknown` in this case, so it's not something you can rely on.

I don't want `SymbolSection` and `SymbolKind` merged. If we were to do anything, it would be to delete `SymbolKind` and require you to find the section kind instead. But as you point out, doing this would lose some information for imported symbols for some file formats, and overall it would be less convenient.

Reply:
(Also, `SymbolKind` covers artificial symbols too, such as file symbols.)

Reply:
It still feels weird that imports get an "undefined" section instead of either the section in which they're defined or the section they are pointing to (like exports)... but I guess okay, I can change the PR, as for now the goal is just to match the unified API rather than provide all possible details.

I do wonder though how we're going to extend them if/when in future we want to collect and distinguish e.g. imported, local and exported globals, tables and memories. I guess we'll need to introduce new symbol kinds, if using `SymbolSection` for imports is not an option?

Reply:
It's not weird at all for ELF/Mach-O/COFF. Their symbols have a section field which has a specific value that means undefined. While they don't explicitly call them imports, an undefined symbol is exactly the same as an import. And I'll repeat my point that changing the semantics for Wasm is not an option.

Adding new symbol kinds is a possibility, but I don't understand enough about what symbols you are planning to add that would need that. None of the symbols added in this PR need it do they? What symbols are there besides code and data?

Reply:
By the way, while I think this PR will be a good change, it's possible you don't need the unified API that this crate provides. This will be the case if the code you are writing will only work for Wasm (and not also ELF/Mach-O/COFF). In that case, also have the option of adding methods to `WasmFile` that are specific to the Wasm functionality you need (or use wasmparser directly).

Reply:
> but I don't understand enough about what symbols you are planning to add that would need that. None of the symbols added in this PR need it do they? What symbols are there besides code and data?

I've mentioned few above - we might want to distinguish memories from tables from globals in the future, as semantically they're different things (even if these semantics are unique to Wasm symbols and not others).

> it's possible you don't need the unified API that this crate provides

I'm aware, and actually already using wasmparser directly for now, but I still think it would be useful to have a unified API for being able to match DWARF information to public symbols via same API across various crates. And, even if I end up keeping separate handlers for Wasm and other formats, such addition might be useful for other consumers of the `object` crate too, hence the upstreaming.

Reply:
I'd glossed over the tables/memories because I don't understand enough about Wasm. I've had a look now. I think it may be better to use `SymbolFlags::Wasm { ... }` for these because this will be something that is specific to Wasm.

Reply:
Hmm yeah I suppose that's another option. I probably won't be able to return to this PR until ~Tuesday, but will update correspondingly as soon as possible.

Reply:
I think we can leave `SymbolFlags` as a future addition (it's not a breaking change). For now I've changed the `section` field as per discussion, so I believe this should be resolved.